### ISSUE TITLE
[Bug]: Can't save optimizer using Checkpointer

### ISSUE THREAD

**Issue by saloelia** [speechbrain/speechbrain#1926]
### Describe the bug

It is not possible to give the checkpointer an optimizer pytorch class in 'recoverables'.

### Expected behaviour

It will be possible to save the optimizer parameters using the Checkpointer class without throwing an error.

**Comment by saloelia** [comment]
**Correction:
It is not possible to give the Checkpointer class an optimizer in 'recoverables' while using the BrainClass for training.

**Comment by Gastron** [comment]
Brain adds the optimizer to the recoverables when you pass it in the opt_class argument: https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/core.py#L854-L869

If you want to add it yourself, you can do so, but the difficulty is that the optimizer takes the optimized parameters as an argument so in the normal SpeechBrain recipe pattern with HyperPyYAML, the optimizer is difficult to construct in YAML, and is easier if it just gets created later by Brain. Some recipes use a pattern where the Brain.init_optimizers method is overridden in the subclass, see e.g. https://github.com/speechbrain/speechbrain/blob/2f86201504550703aacce0df01fbad7a1d9eb011/recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py#L212-L233

**Comment by Adel-Moumen** [comment]
Hi @saloelia,

Any news regarding this issue please? Thanks :-)

**Comment by matthewkperez** [comment]
I experience the same issue as the @saloelia. When I look at my checkpointer.recoverables I see the following objects: ['model', 'lr_annealing', 'normalizer', 'counter', 'brain'].

But no 'optimizer'. The above objects are specified in my yaml so they are expected, but I still experience a reset of my LR when restarting training from a previous checkpoint.

This is after my model has trained for a bit, and I restart training and log asr_brain.checkpointer right after asr_brain initialization.

**Comment by Gastron** [comment]
Could either of you provide a minimal example that reproduces this?

**Comment by Adel-Moumen** [comment]
@Gastron I experienced the same issue when I tried to implement the Lion Optimiser in SpeechBrain. If you are still willing to solve this issue I can share my scripts.

**Comment by Gastron** [comment]
I am happy to have a look at some scripts, though some minimal example that reproduces the problem would be nice.

Also I just re-read this comment:
> This is after my model has trained for a bit, and I restart training and log asr_brain.checkpointer right after asr_brain initialization
I want to add that this is indeed what should be happening - Brain adds the optimizer to the Checkpointer after init_optimizers, which is called at `on_fit_start()`. If this is what the other askers saw as well, then I conclude that Optimizer really only gets added to the checkpointer at fit start, but it will be recovered.

**Comment by Gastron** [comment]
Closing - looks like no bug exists.

We discussed with @Adel-Moumen, and realised that the issue doesn't seem to stem from the Checkpointer. For others, as discussed here, it seems this is mainly a misunderstanding regarding when the Optimizer is initialized and added to the Checkpointer - it happens in the Brain subclass during `on_fit_start()`, which is after the call to `fit()`, naturally.
